### ISSUE TITLE
PCA visualization

### ISSUE THREAD

**Issue by siyuan-mei** [MohammedSB/DINOv2ForMedical#4]
Dear respective authors,

I tried to visualize the first three components of the extracted features in RGB image. However, I can not get the same feature map shown in the paper even if I used the same input figure. For example, the chest X-ray image always contains the body contour. I attach the acquired feature compared to yours.

![img_1](https://github.com/user-attachments/assets/f38e10f9-0b75-4143-a3d6-81c3fabd7638)
![img_1_pca](https://github.com/user-attachments/assets/8ccda186-c5a9-4976-8194-7123cdd220b5)
![img1_pca_rgb](https://github.com/user-attachments/assets/7209a534-a6bd-4c56-8971-21cc30ef66b4)

Is there any wrong here?

**Comment by MohammedSB** [comment]
Can you show me the code you are using to generate your visualization? Keep in mind that you have to pass in multiple X-ray images (4 in our case, shown in the paper), for it to compare feature similarity across these images. We used this issue for our implementation:
https://github.com/facebookresearch/dinov2/issues/23

You can also find a PCA implementation here: DINOv2ForRadiology/dinov2/visualization/pca.ipynb, but I can't remember if we used that, or this huggingface space, which is no longer available for some reason:
https://huggingface.co/spaces/RoundtTble/dinov2-pca

**Comment by wuyouliaoxi** [comment]
thank you for your reply! here are my codes, this time I found the input image needs to be firstly rescaled to [0, 1], and I used vitl_reg. The visualized results get better but still don't reach the performance in the paper.

`dinov2_vitl14_reg = torch.hub.load('facebookresearch/dinov2', 'dinov2_vitl14_reg').cuda()

transform = T.Compose([T.Resize(518, interpolation=InterpolationMode.BICUBIC),
                       T.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
                       ])

features = torch.zeros(4, patch_h * patch_w, feat_dim)
imgs_tensor = torch.zeros(4, 3, 518, 518)
image_folder = '../image_samples/nih_images/'
for i in range(0, 4):
    img_path = image_folder + f'img_{i + 1}.png'
    img = Image.open(img_path)
    img = torch.from_numpy(np.array(img)).unsqueeze(0).repeat(3, 1, 1).float()
    img = (img - img.min()) / (img.max() - img.min())  # Min-max normalization
    imgs_tensor[i] = transform(img)[:3]
with torch.no_grad():
    features_dict = dinov2_vitl14_reg.forward_features(imgs_tensor.cuda())
    features = features_dict['x_norm_patchtokens']
    print(features.shape)
features = features.reshape(4 * patch_h * patch_w, feat_dim).cpu().numpy()
pca = PCA(n_components=3)
pca.fit(features)
pca_features = pca.transform(features)
pca_features = minmax_scale(pca_features)
for i in range(0, 4):
    pca_feature = pca_features[i * patch_h * patch_w: (i + 1) * patch_h * patch_w]
    pca_feature_0 = minmax_scale(pca_feature[:, 0])
    pca_feature_bg = pca_feature_0 < 0.4
    pca_feature_fg = ~pca_feature_bg
    cv2.imwrite(image_folder + f'img{i + 1}_pca_bg.png', pca_feature_bg.reshape(patch_h, patch_w) * 255)
    cv2.imwrite(image_folder + f'img{i + 1}_pca_fg.png', pca_feature_fg.reshape(patch_h, patch_w) * 255)
    pca.fit(pca_feature[pca_feature_fg])
    pca_feature_rem = pca.transform(pca_feature[pca_feature_fg])
    for j in range(3):
        pca_feature_rem[:, j] = minmax_scale(pca_feature_rem[:, j])
    pca_feature_rgb = pca_feature.copy()
    pca_feature_rgb[pca_feature_bg] = 0
    pca_feature_rgb[pca_feature_fg] = pca_feature_rem
    pca_feature_rgb = pca_feature_rgb.reshape(patch_h, patch_w, 3)
    cv2.imwrite(image_folder + f'img{i+1}_pca_rgb.png', pca_feature_rgb * 255, [int(cv2.IMWRITE_PNG_COMPRESSION), 9])
    plt.subplot(2, 2, i+1)
    plt.imshow(pca_feature_rgb)
plt.savefig('features.png')
plt.show()
plt.close()
`

![img1_pca_rgb](https://github.com/user-attachments/assets/ff9d49a9-98bb-464d-a696-e26e7f72334b)

**Comment by siyuan-mei** [comment]
Thank you for your reply! I figured it out.. the reasons are the input image needs to be firstly rescaled to [0, 1] and I used vit_reg

**Comment by MohammedSB** [comment]
Great! Please let me know if you have any other questions.
